transformer_engine/jax/mlp.py

```diff
@@ -3,389 +3,143 @@
 # See LICENSE for license information.
 """JAX MLP modules"""
 
-from typing import List, Tuple
+from typing import List, Tuple, Sequence, Union, Callable
 from functools import partial
 
 import jax
 import jax.numpy as jnp
 from jax.ad_checkpoint import checkpoint_name
 
 from .cpp_extensions import cast_fp8, transpose, cast_transpose
-from .cpp_extensions import gelu as te_gelu
-from .cpp_extensions import gelu_fp8, dgelu, dgelu_dbias_cast_transpose
+from .cpp_extensions import gelu, gelu_fp8
+from .cpp_extensions import dgelu, dgelu_dbias_cast_transpose
 from .cpp_extensions import gated_gelu, gated_gelu_fp8
 from .cpp_extensions import dgated_gelu, dgated_gelu_cast_transpose
+from .cpp_extensions import silu, silu_fp8
+from .cpp_extensions import dsilu, dsilu_dbias_cast_transpose
+from .cpp_extensions import gated_silu, gated_silu_fp8
+from .cpp_extensions import dgated_silu, dgated_silu_cast_transpose
 from .cpp_extensions import rmsnorm_fwd_fp8, rmsnorm_bwd
 from .cpp_extensions import layernorm_fwd_fp8, layernorm_bwd
-from .dot import fp8_dot_impl, get_precision_of_fp8_dot, quantize
+from .dot import fp8_dot_impl, get_precision_of_fp8_dot, quantize, dequantize
 from .layernorm import canonicalize_layernorm_type
 from .fp8 import FP8Helper, FP8MetaPackage
 from .sharding import with_sharding_constraint_by_logical_axes
 
 
-def gelu(x: jnp.ndarray):
+activation_dict = {
+    ('gelu',): { 'fwd': gelu,
+                "bwd": dgelu },
+    ('gelu', 'linear'): { 'fwd': gated_gelu,
+                         'bwd': dgated_gelu },
+    ('silu',): { 'fwd': silu,
+                "bwd": dsilu },
+    ('silu', 'linear'): { 'fwd': gated_silu,
+                         'bwd': dgated_silu }
+}
+
+activation_fp8_dict = {
+    ('gelu',): { 'fwd': gelu_fp8,
+                'bwd': dgelu_dbias_cast_transpose },
+    ('gelu', 'linear'): { 'fwd': gated_gelu_fp8,
+                          'bwd': dgated_gelu_cast_transpose },
```

Comment:
@phu0ngng and I already talked about this offline as a future PR but just dropping a note here for others.

We should ideally have `fused_layernorm_fp8_mlp()` support biases for every activation, and the activations that do not have `d<act>_dbias_cast_transpose()` implemented should do `d<act>()` + `cast_tranpose_dbias()`.

Biases for these activations are supported in `te.flax.LayerNormMLP` through unfused composition, but we do have `cast_transpose_dbias()` available to us in TE and we should make use of it here to match the TE/PyTorch implementation.

Reply:
Hi, this was addressed in #765. So I marked it as resolved here.